### PR TITLE
feat(metrics): Add new sliding-window rate limiter [INGEST-1380]

### PR DESCRIPTION
**Pull request by untitaker**
Add a sliding window rate limiter to Sentry that will eventually be used by the string indexer to throttle metric tag/name writes to postgres.

The implementation does not allow for dynamic configuration. It's expected that this is all handled inside of the string indexer, who is also responsible for generating a ton of quota objects per batch.

The rough idea is that per message batch in the string indexer, we will construct one `RequestedQuota` per organization seen in the batch, with a prefix of `prefix=f"string-indexer:o:{org_id}"`.

Each `RequestedQuota` will have 2x2 `Quota` entries:

* one dimension is for the number of windows: one for the per-second (per-10-second?) limit, and one for a 24h limit
* another dimension for organization vs global limit: the global limit's Quota will have a `prefix_override` set that does not include the org ID, while the per-organization quota will inherit the `prefix` from the `RequestedQuota`.

Those quotas are then sent to `check_within_quotas`, the indexer gets back the allowed number of strings per organization and globally, heeds both the global and per-org limit, then consumes the appropriate quotas at the end of the batch.

A future implementation of this rate limiter could implement the entire flow as lua script. Both checking and consuming the quota could be done in one step, atomically, since the "request format" is complex enough to express everything we want to limit by (there's no post-processing of quota grants in the string indexer)

### CODE REVIEW

**Review by evanh — COMMENTED** [review]
This is pretty cool, nicely done.

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:60]: Is this something you'd want to enforce?

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:27]: I would suggest appending units to the variable names such that it's easier for the user. E.g`window_seconds`, `granularity_seconds`.

When looking at the tests I had to come back here to figure out what `window` and `granularity` were

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:145]: this function is a little difficult to follow along with. The code is totally fine but it's hard to map what is going on in the code to what is going on in redis. 

An example of how the data can look like in the comments can make it much easier to follow along with what is happening. This was done for the [snuba rate limiter](https://github.com/getsentry/snuba/blob/master/snuba/state/rate_limit.py#L143-L170).

You can also provide the intention behind code in [human language](https://github.com/getsentry/sentry/blob/master/src/sentry/scripts/ratelimits/api_limiter.lua#L39-L40) so that the intent is clear without the reader having to run the code in their head to figure it out.

**Review by volokluev — COMMENTED** [review]
It mostly looks fine to me but I think you definitely want to get a review from the users of this module because they actually know how they will use it better and whether it fits their needs

File: src/sentry/ratelimits/sliding_windows.py
Comment (fpacifici) [inline src/sentry/ratelimits/sliding_windows.py:122]: How do we deal with redis failovers in the sentry code base? If we keep a reference to the cluster as an attribute and redis has a failover, do we let this fail and then re-establish the connection ?

File: src/sentry/ratelimits/sliding_windows.py
Comment (fpacifici) [inline src/sentry/ratelimits/sliding_windows.py:135]: Was there any reason to have a key for each granule vss having a set per `prefif:window`  and have each granule represented as an element of the set ?

File: src/sentry/ratelimits/sliding_windows.py
Comment (onewland) [inline src/sentry/ratelimits/sliding_windows.py:51]: kind of a generic python question -- is `int(num/div)` preferable to `num // div` for some reason?

File: src/sentry/ratelimits/sliding_windows.py
Comment (onewland) [reply src/sentry/ratelimits/sliding_windows.py:85]: looks like it defaults to current time in the redis implementation -- I think that should probably be in the doc comment (e.g. "implementors of `SlidingWindowRateLimiter` should assume that `None` time should be checked against the current timestamp")

File: src/sentry/ratelimits/sliding_windows.py
Comment (onewland) [reply src/sentry/ratelimits/sliding_windows.py:101]: Following up on Filippo's statement, I think given that this is an interface/abstract class we should consider *not* surfacing this implementation detail (the existence of `use_quotas` and `check_quotas`) as (1) it's specific to the Redis sliding window rate limiter, and (2) I actually don't particularly see the value of a separate check and use for this use case. I understand the example provided but I'd rather err on the side of slightly too restrictive than slightly too permissive.

**Review by onewland — COMMENTED** [review]
I'm mostly fine with this but I think we should consider restricting the rate limiter interface even more

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:60]: whoops, I added the validation for this. The idea was to design and restrict the rate limiter interface in such a way that it can be transparently refactored to use a Lua script internally. In that case we need to define our own sharding inside of the rate limiter, and users who define their own sharding would experience a breaking change.

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:135]: because keys have TTL and set elements don't. yes this increases memory usage due to duplicated keys, but i don't think it's a big problem, we also have the same setup in relay.

I see that the concurrent rate limiter has its own mechanism for implementing TTL inside of sets, but there we need to determine set length/cardinality, and here we don't

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:122]: I had to check because I don't know, but my impression is that we deal with it poorly.

Apparently we have the [RetryingRedisCluster](https://github.com/getsentry/sentry/blob/834adc3c1dc5ac1b2035ec61236e0716792f5c6d/src/sentry/utils/redis.py#L70) which reestablishes the connections transparently on some errors, but i don't know if that is enough.

We also recently added [FailoverRedis](https://github.com/getsentry/sentry/blob/834adc3c1dc5ac1b2035ec61236e0716792f5c6d/src/sentry/utils/redis.py#L92) which indicates that previously we just didn't deal with failovers? This class is only used in a single redis cluster (configured in getsentry) @beezz do you know more about this?

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:101]: I added a bunch of doccomments that hopefully clarify what i am thinking by having this separate interface

@onewland I am not sure if overspending is good during incident recovery (see comment about crashloops). We also had a discussion about how to recover during backlogs, where I suppose the idea was to just have very long windows.

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:82]: both 😅 they would be a prefix within each prefix:

* a request with `prefix=f"release-health:{org_id}"` for the per-org limit in release health
  * two quotas with `prefix_override=None` for a long and a short window
  * within the same request, we have another two quotas with `prefix_override=f"release-health:global"` for the global quota

* a request with `prefix=f"mep:{org_id}" for the per-org limit in MEP
  * again two quotas with `prefix_override=None`
  * again two extra quotas for the global quotas, `prefix_override=f"mep:global"`

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:106]: added it

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py]: nope i forgot how to write python

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py:136]: However if we want separate redis clusters per usecase I think you'll currently have to just instantiate the service twice.

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:161]: can you document the return type? defining a `Timestamp = int` type will make this signature easier to understand

File: tests/sentry/ratelimits/test_sliding_windows.py
Comment (volokluev) [inline tests/sentry/ratelimits/test_sliding_windows.py:48]: food for thought:

I know that `check_and_use_quotas` just calls `check_within_quotas` and then `use_quotas` but it doesn't seem like it's going to be used this way which makes me wonder if the tests need to change or if the interface needs to be different

e.g.

```python3
with RedisSlidingWindowRateLimiter.check_and_use_quotas(....): # check_within_quotas
    # do stuff
# use quotas at exit
```

I won't be using this library so I can't say for sure but just something to think about

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:51]: nit

if you make your limits smaller numbers (e.g. 4, 2 respectively) then the ellipses become unnecessary and the explanation is possibly clearer because everything fits on the page (and into the reader's head)

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:65]: I think it's worth explaining the non-empty case as that is more likely to be the usual state of things. Namely that:

* the used quota for the window is calculated by summing the granules
* that the smallest remaining quota will be the one that is granted

**Review by volokluev — APPROVED** [review]
I approve this PR from the following point of view:

* The code is understandable 
* Tradeoffs are documented and clear
* As far as I can tell there are no obvious bugs that I can catch as code reviewer

I left a few non-blocking comments on how it could be made clearer.

I do not have context on the greater project or the use of this code and therefore should not be the sole approver of this. However, as far as I'm concerned I don't think any further review from me is necessary unless there are some significant changes.

File: tests/sentry/ratelimits/test_sliding_windows.py
Comment (untitaker) [reply tests/sentry/ratelimits/test_sliding_windows.py:48]: `check_and_use_quotas` is only going to be used in tests right now but I put it on the baseclass anyway just because I _could_ imagine it _could_ be useful in other scenarios

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py]: I updated the numbers, please check. I updated the granularity and window because the number of keys is defined by `window/granularity`

File: src/sentry/ratelimits/sliding_windows.py
Comment (untitaker) [reply src/sentry/ratelimits/sliding_windows.py]: ok, i added a little bit more info

**Review by onewland — APPROVED** [review]

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:33]:
```suggestion
                # window should be. This one configures per-10-second granularity
                # to make the example simpler
```

File: src/sentry/ratelimits/sliding_windows.py
Comment (volokluev) [inline src/sentry/ratelimits/sliding_windows.py:60]: "After one minute" and "at time 902" are opposing statements are they not? (or at least confusing. 
```suggestion
Another request for the same org happens at time `902`.
```

**Review by volokluev — COMMENTED** [review]
A few small comments, but LGTM

**Pull request merged**
